### PR TITLE
Added Windows (win32) support

### PR DESCRIPTION
**Pull request by jnstahl**
TODO:
-allow the query to run repeatedly until explicitly stopped
-allow multiple queries in parallel (?)

### CODE REVIEW

**Comment by jnstahl** [comment]
OK so this last commit should meet the basic requirements. I modified the sample app so that it will look for generic _http._tcp servers in the local network. It seems to run correctly on windows

**Comment by Nimrodda** [comment]
Could you please rebase on latest master branch?

File: windows/mdns.h [hunk 1]
```diff
@@ -0,0 +1,1535 @@
+/* mdns.h  -  mDNS/DNS-SD library  -  Public Domain  -  2017 Mattias Jansson
```
Comment (Nimrodda) [inline windows/mdns.h:1]: I'm not familiar with Windows development, but why are we including a library inside the plugin codebase?
I'd prefer that we either depend on official Windows API or use a dependency management system to depend on this library.

File: windows/flutter_nsd_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+      version_stream << "10+";
+    }
+    else if (IsWindows8OrGreater()) {
+      version_stream << "8";
+    }
+    else if (IsWindows7OrGreater()) {
+      version_stream << "7";
+    }
+    result->Success(flutter::EncodableValue(version_stream.str()));
+  }
```
Comment (Nimrodda) [inline windows/flutter_nsd_plugin.cpp]: This should be removed as the plugin doesn't support such method. Only `startDiscovery` and `stopDiscovery`.

File: windows/flutter_nsd_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    if (currentRequest != NULL) {
+      auto req = currentRequest;
+      currentRequest = NULL;
+      req->stop();
+
+    }
+    result->Success();
+  }
+  else {
+    result->Success();
```
Comment (Nimrodda) [inline windows/flutter_nsd_plugin.cpp]: I'd prefer to return a failure here in case an unsupported method was called

File: windows/mdns.h (hunk 1 above)
Comment (Nimrodda) [reply windows/mdns.h:1]: Maybe via [Conan](https://conan.io)? I see it is published there https://github.com/mjansson/mdns#using-with-cmake-conan-or-vcpkg

File: flutter_nsd.iml [hunk 4]
```diff
@@ -23,7 +23,6 @@
       <excludeFolder url="file://$MODULE_DIR$/example/ios/.symlinks/plugins/flutter_nsd/build" />
     </content>
     <orderEntry type="sourceFolder" forTests="false" />
-    <orderEntry type="library" name="Dart Packages" level="project" />
```
Comment (Nimrodda) [inline flutter_nsd.iml:26]: Not sure why this was removed

**Review by Nimrodda — COMMENTED** [review]
Which IDE can I use to build the Windows code?

**Comment by Nimrodda** [comment]
Thanks for contributing. Looks good overall. I made a few comments though that should be addressed. Thanks!

File: windows/mdns.h (hunk 1 above)
Comment (jnstahl) [reply windows/mdns.h:1]: > Maybe via [Conan](https://conan.io)? I see it is published there https://github.com/mjansson/mdns#using-with-cmake-conan-or-vcpkg

Unfortunately, windows has no built-in support for mdns. So we have to rely on 3rd party libraries. There are several available. I chose this one because it was very self-contained (just a single .h file) and seemed relatively stable.

I agree it would be better to include a dependency rather than copy/paste. I just did not know how to accomplish that so I wanted to see if it would work in the first place.

**Comment by Nimrodda** [comment]
No worries, indeed it's the fastest way to try it out :)
I did a little research myself on Flutter and C++ dependencies. There
doesn't seem to be anything built-in.
So what I suggest is that we get it via Conan and just mention it in the
readme file as a pre-build step that needs to be run for Windows to work.

File: windows/flutter_nsd_plugin.cpp (hunk 2 above)
Comment (jnstahl) [reply windows/flutter_nsd_plugin.cpp]: Yes. This is added automatically by the flutter create command. It is even added automatically for the other platforms such as iOS or Android. So it seems to be sort of a standard API. You may want to consider adding it to the other platforms:

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

 (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else {
    result(FlutterMethodNotImplemented);
  }


this code was taken from an empty plugin that I just generated to verify tis.

File: windows/flutter_nsd_plugin.cpp (hunk 3 above)
Comment (jnstahl) [reply windows/flutter_nsd_plugin.cpp]: ugh. copy/paste error

File: flutter_nsd.iml (hunk 4 above)
Comment (jnstahl) [reply flutter_nsd.iml:26]: Not intentionally. this is an ide config file.

**Comment by jnstahl** [comment]
> Which IDE can I use to build the Windows code?

https://docs.flutter.dev/desktop

For Windows desktop development, you need the following in addition to the Flutter SDK:

Visual Studio 2019 (not to be confused with Visual Studio Code). For Win32 you need the “Desktop development with C++” workload installed, including all of its default components. For UWP you need the “Universal Windows Platform development” workload installed, with the optional UWP C++ tools.

**Comment by jnstahl** [comment]
> No worries, indeed it's the fastest way to try it out :) I did a little research myself on Flutter and C++ dependencies. There doesn't seem to be anything built-in. So what I suggest is that we get it via Conan and just mention it in the readme file as a pre-build step that needs to be run for Windows to work.

How does this conan thing work? Is there no way just to reference the .h file within github?

**Comment by jnstahl** [comment]
There are a couple differences in the results sent by the plugin.
In your Android code you return:

  val hostname = this?.host?.canonicalHostName

which may be a dns name or an ip address as fallback.

I just added some code in the latest commit which will somehow emulate this.. But I also include ipv4 and ipv6 addresses anyway. This is currently ignored at the Dart level.

File: windows/flutter_nsd_plugin.cpp (hunk 2 above)
Comment (jnstahl) [reply windows/flutter_nsd_plugin.cpp]: If you want I can add support for these calls to the other platforms

**Comment by Nimrodda** [comment]
yes, I know. It's just an example that they provide for a fast start 😅

File: .idea/runConfigurations.xml [hunk 5]
```diff
@@ -1,10 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Nimrodda) [inline .idea/runConfigurations.xml:1]: Could you please revert the deletion of this file?

File: .gitmodules [hunk 6]
```diff
@@ -0,0 +1,3 @@
+[submodule "windows/mdns"]
```
Comment (Nimrodda) [inline .gitmodules]: Did you try with Conan? I have really bad experience with Git submodules

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,8 @@
+## 1.4.0-alpha01
+
+Add support for Windows using the https://github.com/mjansson/mdns library
+
+
```
Comment (Nimrodda) [inline CHANGELOG.md]: Let's add it under 1.3.0 release :)

File: pubspec.yaml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 name: flutter_nsd
-description: A Flutter plugin that enables Network Service Discovery on various platforms via their native APIs.
-version: 1.3.0
+description: Flutter Network Service Discovery plugin
+version: 1.4.0-alpha01
```
Comment (Nimrodda) [inline pubspec.yaml]: Please don't make any changes to this file. It is good as is.

File: windows/flutter_nsd_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+    auto serviceType = arguments->find(flutter::EncodableValue("serviceType"));
+    std::string type = "";
+    if (serviceType != arguments->end()) {
+      type = std::get<std::string>(serviceType->second);
+    }
+
+    if (currentRequest != NULL) {
+      auto req = currentRequest;
+      currentRequest = NULL; // request will delete itself after the thread exits, we can detach the pointer here
+      req->stop(); // will stop the background thread after the next timeout; returns immediately
```
Comment (Nimrodda) [inline windows/flutter_nsd_plugin.cpp:263]: What happens if `start()` is called before the next timeout? Would we have 2 scans ongoing?

File: windows/mdns_impl.c [hunk 10]
```diff
@@ -0,0 +1,525 @@
+
+#include <stdio.h>
```
Comment (Nimrodda) [inline windows/mdns_impl.c:2]: The MDNS library is still included in the PR :(

**Review by Nimrodda — CHANGES_REQUESTED** [review]

File: windows/flutter_nsd_plugin.cpp (hunk 9 above)
Comment (jnstahl) [reply windows/flutter_nsd_plugin.cpp:263]: Yes - it would be very difficult to preempt the running query as it may be stuck in a select() call
It should be OK - the only thing is that you may receive results from both queries until the first one times out

Comment (Nimrodda) [reply windows/flutter_nsd_plugin.cpp:263]: Sounds good to me 👍

File: windows/mdns_impl.c (hunk 10 above)
Comment (jnstahl) [reply windows/mdns_impl.c:2]: the way the library is set up is that the mdns.h file has all the implementation. The mdns.c files is an example command line app demonstrating how to use it. It includes also code for MDNS server and other aspects. The mdns_impl.c was generated by selective copy/paste and modification of the sample code to implement the specific use case. So this file needs to remain.

File: .idea/runConfigurations.xml (hunk 5 above)
Comment (jnstahl) [reply .idea/runConfigurations.xml:1]: done with the next commit

File: .gitmodules (hunk 6 above)
Comment (jnstahl) [reply .gitmodules]: I spent some time looking into it - it seems to be more of a CID tool which generates binaries? I could not figure out how to integrate it into either Android Studio or Visual Studio? all we really need is to pull the mdns.h file from the repo

File: CHANGELOG.md (hunk 7 above)
Comment (jnstahl) [reply CHANGELOG.md]: fixed in the next commit

File: pubspec.yaml (hunk 8 above)
Comment (jnstahl) [reply pubspec.yaml]: reverted in the last commit

**Comment by Nimrodda** [comment]
> > No worries, indeed it's the fastest way to try it out :) I did a little research myself on Flutter and C++ dependencies. There doesn't seem to be anything built-in. So what I suggest is that we get it via Conan and just mention it in the readme file as a pre-build step that needs to be run for Windows to work.
> 
> How does this conan thing work? Is there no way just to reference the .h file within github?

You can find more info about it here: https://docs.conan.io/en/latest/getting_started.html
But shortly, it's like Maven

File: .gitmodules (hunk 6 above)
Comment (Nimrodda) [reply .gitmodules]: I haven't used it myself, but just briefly reading the getting started guide, it looked to me like Maven so it's just a command line tool.
But well, I can try to integrate it at some point.

File: windows/mdns_impl.c (hunk 10 above)
Comment (Nimrodda) [reply windows/mdns_impl.c:2]: Ahh ok. That's clear now 👍

**Comment by Nimrodda** [comment]
> > Which IDE can I use to build the Windows code?
> 
> https://docs.flutter.dev/desktop
> 
> For Windows desktop development, you need the following in addition to the Flutter SDK:
> 
> Visual Studio 2019 (not to be confused with Visual Studio Code). For Win32 you need the “Desktop development with C++” workload installed, including all of its default components. For UWP you need the “Universal Windows Platform development” workload installed, with the optional UWP C++ tools.

I think it would be a good idea to include this info in the README.md file, under Windows section you added.
It would be really great if you could include links as well to the stuff that one needs to download, like this "Desktop development with C++", etc.

I think once we have that, we're good to merge. Thanks a lot!

File: .gitmodules (hunk 6 above)
Comment (jnstahl) [reply .gitmodules]: I think this requires further study. How does the pub.dev deployment work? From what I can tell is that when you type flutter build windows, it creates a visual studio project and copies all the sources into it, including the sources out of the plugin's windows folder. So all the sources need to be in there. We certainly do not want that the plugin user needs to install the conan client just to get all the sources.

Comment (Nimrodda) [reply .gitmodules]: Good point. I don't know how pub.dev deployment works. We'll have to try and see :)

Comment (jnstahl) [reply .gitmodules]: OK the way it works right now with the submodules is that if you clone the repo, the mdns sources are not included. One needs to run git submodule update to pull in the sources.

I did a flutter pub publish --dry-run and this is what would be published from the windows folder:

'-- windows
    |-- CMakeLists.txt
    |-- flutter_nsd_plugin.cpp
    |-- include
    |   '-- flutter_nsd
    |       '-- flutter_nsd_plugin.h
    |-- mdns
    |   |-- CHANGELOG
    |   |-- CMakeLists.txt
    |   |-- LICENSE
    |   |-- README.md
    |   |-- cmake
    |   |   '-- Config.cmake.in
    |   |-- mdns.c
    |   '-- mdns.h
    |-- mdns_impl.c
    '-- mdns_impl.h

as you can see, the entire mdns repo will be included, not just the mdns.h file. Not ideal. But the only way to fix that would be to just manually copy/paste the file

the guidelines say:
Your package should depend only on hosted dependencies (from the default pub package server) and SDK dependencies (sdk: flutter). These restrictions ensure that dependencies of your packages cannot become unavailable in the future. So we need to include the mdns.h file in the package, and not just a dependency.

**Comment by jnstahl** [comment]
> 

I included a description of the IDE requriements. However anyone building an app for Windows should have this already, even before they they run into this dependency when compiling the plugin

File: .gitmodules (hunk 6 above)
Comment (jnstahl) [reply .gitmodules]: So I would suggest that I remove the sub-package dependency, then copy the mdns.h file manually (just as I did in the first version), then add a README.md into that folder that tells where that file came from and how to update it if needed.

Comment (Nimrodda) [reply .gitmodules]: I had a look into another Flutter plugin with Windows support and I see that they're including the library dependencies in binary format (either lib/dll). Actually not only that they include dependencies in binary format, but also the C++ code itself. The only visible code are the Flutter plugin cpp and h files. You can see it here for example: https://github.com/dynamsoftlabs/flutter_barcode_sdk/tree/main/windows

So basically, we'll need all the C++ code to be in binary format and their header files and `flutter_nsd_plugin.cpp` + `flutter_nsd_plugin.h`.

Comment (Nimrodda) [reply .gitmodules]: Another Flutter plugin with Windows support that follows the same pattern: binaries + headers + own plugin cpp/h. https://github.com/jerson/flutter-openpgp/tree/master/windows

**Comment by Nimrodda** [comment]
> > 
> 
> I included a description of the IDE requriements. However anyone building an app for Windows should have this already, even before they they run into this dependency when compiling the plugin

Sure thing, but I will need to test this before I publish and I don't have my system set up yet for Windows app building :)

File: .gitmodules (hunk 6 above)
Comment (jnstahl) [reply .gitmodules]: Yes I can see that. However it would depend on how these libraries are distributed originally. The mdns lib is just in a .h file so it may be a bit tricky to turn that into a binary. One could compile the mdns_impl.c however that is part of the plugin source. either way I tried building a project with the sources included and it compiled just fine.

**Comment by jnstahl** [comment]
> > > 
> > 
> > 
> > I included a description of the IDE requriements. However anyone building an app for Windows should have this already, even before they they run into this dependency when compiling the plugin
> 
> Sure thing, but I will need to test this before I publish and I don't have my system set up yet for Windows app building :)

for testing you can link my fork into your flutter project like this:


 flutter_nsd:
    git:
      url: https://github.com/jnstahl/flutter_nsd
      ref: windows

File: .gitmodules (hunk 6 above)
Comment (Nimrodda) [reply .gitmodules]: Can we rename the mdns.h to .cpp and extract a header file for it so that we can compile it and include it as a binary?

Comment (jnstahl) [reply .gitmodules]: I don't think it is a rule to include binaries. This may apply if those libraries are not available as source. But in this case where the lib is distributed as source it makes more sense to include the source. The guidelines only say that everything which is required to build the plugin should be included in the package so that it does not rely on external repos which may not be there in the future.
And this mdns.h is explicitly labeled as public domain

File: windows/mdns_impl.c [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+    sock_addr.sin_family = AF_INET;
+#ifdef _WIN32
+    sock_addr.sin_addr = in4addr_any;
+#else
+    sock_addr.sin_addr.s_addr = INADDR_ANY;
+#endif
+    sock_addr.sin_port = htons(MDNS_PORT);
+#ifdef __APPLE__
+    sock_addr.sin_len = sizeof(struct sockaddr_in);
+#endif
```
Comment (Nimrodda) [inline windows/mdns_impl.c:419]: This file has many compiler directives for other platforms than Windows. I think it would be good to remove what we don't need.

File: .gitmodules (hunk 6 above)
Comment (Nimrodda) [reply .gitmodules]: It's true that it's not a rule, but it's also not part of the plugin's source code that we maintain therefore I believe it shouldn't be part of the plugin's Git repository. Maybe I'm just used to Android dev too much :)
Ideally, I'd prefer to have a dependency management system like Cocoapods for MacOS/iOS and Gradle for Android, but looks like Flutter doesn't support any out of the box.

It would be great if we could only include the binaries so that there's a clear separation between plugin code and library code,  but I understand that it might require extra time to get working. If you prefer not to spend more time on this, then we can merge it and I'll look into it at some point. Thanks!

**Review by Nimrodda — APPROVED** [review]

**Comment by Nimrodda** [comment]
It would be great if you have the time for a follow up PR to address [removal of non-win32 code](https://github.com/Nimrodda/flutter_nsd/pull/16#pullrequestreview-813388999) and [bundle binaries instead of actual code for mdns library](https://github.com/Nimrodda/flutter_nsd/pull/16#discussion_r755014519).

Thanks for your contribution!
